### ISSUE TITLE
no-extra-bind should flag arrow functions using 'this'

### ISSUE THREAD

**Issue by solmsted** [eslint/eslint#3672]
Arrow functions bind `this` lexically.  When an arrow function is used with `Function.prototype.apply`, `Function.prototype.bind`, `Function.prototype.call`, or `Reflect.apply` the thisArgument is silently ignored.  These functions are still useful for specifying arguments to the arrow function but they can create confusing errors if the new value of `this` is expected to be applied.  I propose a new rule be added to warn when trying to set a new value for `this` in an arrow function.

I'm not sure exactly how to define such a rule.  My initial thought is to warn whenever a non-null value is used as the thisArgument for an arrow function.  By always passing `null` as the thisArgument, no one would ever be confused as to why the function wasn't using their custom object as `this`.  On the other hand, someone might expect `this` to be set to `null` within the arrow function and still cause confusion.

**Comment by eslintbot** [comment]
Thanks for the issue! We get a lot of issues, so this message is automatically posted to each one to help you check that you've included all of the information we need to help you.

Reporting a bug? Please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
1. The source code that caused the problem
1. The configuration you're using (for the rule or your entire config file)
1. The actual ESLint output complete with line numbers

Requesting a new rule? Please be sure to include:

1. The use case for the rule - what is it trying to prevent or flag?
1. Whether the rule is trying to prevent an error or is purely stylistic
1. Why you believe this rule is generic enough to be included

Requesting a feature? Please be sure to include:

1. The problem you want to solve (don't mention the solution)
1. Your take on the correct solution to problem

Including this information in your issue helps us to triage it and get you a response as quickly as possible.

Thanks!

**Comment by nzakas** [comment]
We already have a rule that does that: http://eslint.org/docs/rules/no-extra-bind

The documentation should be updated, as there are no arrow functions, but the rule definitely flags arrow functions.

**Comment by solmsted** [comment]
I could be mistaken but if I understand you correctly, I feel like the implementation of `no-extra-bind` is incomplete.  Here are some tests I ran.  All of these used version 1.3.1 and this .eslintrc:
```
{
    "ecmaFeatures": {
        "modules": true
    },
    "env": {
        "es6": true
    },
    "root": true,
    "rules": {
        "no-extra-bind": 2
    }
}
```

The basic check as described in the docs warns as expected:
```
const someFunction = function () {}.bind('something');
```
`  1:22  error  The function binding is unnecessary  no-extra-bind`

As you pointed out, it warns as expected with arrow functions too:
```
const someFunction = (() => {}).bind('something');
```
`  1:22  error  The function binding is unnecessary  no-extra-bind`

It does not warn if the arrow function's body contains `this`:
```
const someFunction = (() => {return this;}).bind('something');
```
No warning.  My proposal was intended to address this.  It seems like `someFunction` should always return `'something'` but it will instead return whatever the `this` value happens to be in the parent scope.

I discovered that there is no warning if you call bind on a variable instead of directly on the function expression:
```
const someFunction = function () {},
    someBoundFunction = someFunction.bind('something');
```
No warning.  This wasn't part of my original proposal but I feel like this is a more common pattern and it should generate a `no-extra-bind` warning.  I realize it's not always possible to statically determine the value of a variable but it should be possible in some cases.

Using `Function.prototype.call`:
```
const someValue = function (arg) {return arg;}.call('something', 'someValue');
```
No warning.

Using `Function.prototype.apply`:
```
const someValue = function (arg) {return arg;}.apply('something', [
    'someValue'
]);
```
No warning.

Using `Reflect.apply`:
```
const someValue = Reflect.apply(function (arg) {return arg;}, 'something', [
    'someValue'
]);
```
No warning.

So in summary, I'd like to request a few enhancements.
* Instead of just checking `Function.prototype.bind`, check for all mechanisms that could change the `this` value.
    * There's a dilemma I mentioned originally.  Other functions like `Function.prototype.call` are still perfectly valid for specifying arguments so warnings should only be generated when obviously attempting to set the thisArgument.
* Instead of just checking function expressions also check for extra bindings when the function has been saved to a variable (where possible)
* Instead of just checking functions that don't use `this` also check for arrow functions that contain `this` which can not be rebound.

**Comment by nzakas** [comment]
> I realize it's not always possible to statically determine the value of a variable but it should be possible in some cases.

We don't ever do this, it's too dangerous to check in some cases but not others. It also sets expectations poorly for how ESLint works.

> Instead of just checking Function.prototype.bind, check for all mechanisms that could change the this value.

If we did this, it would be a separate rule. That said, I don't see this as the same as `bind`, so I'm not sure such a rule would be useful. Unlike `bind`, the other methods don't create a new function so there's less occasion for trouble. 

> Instead of just checking function expressions also check for extra bindings when the function has been saved to a variable (where possible)

As already mentioned, we do not track variable values, so we don't do these sorts of checks.

> Instead of just checking functions that don't use this also check for arrow functions that contain this which can not be rebound.

I think this is an actual bug we should fix.

---

Keep in mind, you can always create your own rules that do what you want even if we don't want it in core, so you have options for achieving what you want.

**Comment by solmsted** [comment]
Sounds good.  Thank you for looking into this!

**Comment by nzakas** [comment]
Working on this.
